packages/ampd-sdk/src/grpc/connection_manager.rs

```diff
@@ -0,0 +1,257 @@
+            connection_receiver: state_receiver,
+            message_sender: msg_sender,
+        };
+
+        let manager = Self {
+            url,
+            connection_state: state_sender,
+            message_receiver: msg_receiver,
+            retry_policy: DEFAULT_RETRY_POLICY,
+        };
+
+        Ok((manager, handle))
+    }
+
+    pub async fn run(mut self) -> Result<(), Error> {
+        if let Err(e) = self.connect().await {
+            warn!(err = ?e, "initial connection failed, will retry on client requests");
+        }
+
+        while let Some(message) = self.message_receiver.recv().await {
+            match message {
+                ClientMessage::ConnectionFailed(details) => {
+                    warn!("client reported connection failure: {}", details);
+                    self.handle_connection_failure().await?
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn connect(&self) -> Result<(), Error> {
+        let endpoint: transport::Endpoint = self.url.as_str().parse().into_report()?;
+        let endpoint = endpoint
+            .connect_timeout(DEFAULT_INITIAL_TIMEOUT)
+            .timeout(DEFAULT_RPC_TIMEOUT)
+            .keep_alive_timeout(KEEPALIVE_TIMEOUT)
+            .keep_alive_while_idle(KEEPALIVE_WHILE_IDLE)
+            .http2_keep_alive_interval(KEEPALIVE_TIME);
+
+        match endpoint.connect().await {
+            Ok(channel) => {
+                self.connection_state
+                    .send(ConnectionState::Connected(channel))
+                    .into_report()?;
+                Ok(())
+            }
+            Err(status) => {
+                let status_error = Err(status).into_report();
+                if self
+                    .connection_state
+                    .send(ConnectionState::Disconnected)
+                    .is_err()
+                {
+                    return status_error
+                        .attach_printable("connection failed and could not notify clients");
+                };
+                status_error
+            }
+        }
+    }
+
+    async fn handle_connection_failure(&self) -> Result<(), Error> {
+        self.connection_state
+            .send(ConnectionState::Reconnecting)
+            .into_report()?;
+
+        match with_retry(|| self.connect(), self.retry_policy).await {
+            Ok(()) => {
+                info!("successfully reconnected after failure");
+                Ok(())
+            }
+            Err(error_report) => {
+                error!(err = ?error_report, "failed to reconnect after max attempts");
+                let retry_error = Err(error_report);
+                if self
+                    .connection_state
+                    .send(ConnectionState::Disconnected)
+                    .is_err()
+                {
+                    return retry_error
+                        .attach_printable("connection failed and could not notify clients");
+                };
```

Comment:
I believe this can only fail if the receivers are either down or there is too much back pressure. This seems like something that should just be a warning log and then suppressed.